ext/src/ruby_api/params.rs

```diff
@@ -1,30 +1,41 @@
 use super::convert::ToWasmVal;
 use magnus::{exception::arg_error, Error, ExceptionClass, Value};
+use static_assertions::assert_eq_size;
 use wasmtime::{FuncType, ValType};
 
-#[derive(Debug)]
-struct Param<'a> {
-    index: usize,
-    ty: ValType,
-    val: &'a Value,
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+struct Param {
+    val: Value,
+    index: u32,
+    ty: ValTypeCopy,
 }
+// Keep `Param` small so copying it to the stack is cheap, typically anything
+// less than 3usize is good
+assert_eq_size!(Param, [u64; 2]);
 
-impl<'a> Param<'a> {
-    pub fn new(index: usize, ty: ValType, val: &'a Value) -> Self {
-        Self { index, ty, val }
+impl Param {
+    pub fn new(index: u32, ty: ValType, val: Value) -> Self {
+        Self {
+            index,
+            ty: ty.into(),
+            val,
+        }
     }
 
-    fn to_wasmtime_val(&self) -> Result<wasmtime::Val, Error> {
-        self.val.to_wasm_val(&self.ty).map_err(|error| match error {
-            Error::Error(class, msg) => {
-                Error::new(class, format!("{} (param index {}) ", msg, self.index))
-            }
-            Error::Exception(exception) => Error::new(
-                ExceptionClass::from_value(exception.class().into()).unwrap_or_else(arg_error),
-                format!("{} (param index {}) ", exception, self.index),
-            ),
-            _ => error,
-        })
+    fn to_wasmtime_val(self) -> Result<wasmtime::Val, Error> {
+        self.val
+            .to_wasm_val(self.ty.into())
+            .map_err(|error| match error {
+                Error::Error(class, msg) => {
+                    Error::new(class, format!("{} (param index {}) ", msg, self.index))
```

Comment:
For symmetry, should we update this message too?
```suggestion
                    Error::new(class, format!("{} (param at index {}) ", msg, self.index))
```